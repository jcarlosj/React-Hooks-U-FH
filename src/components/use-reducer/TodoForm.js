import React from 'react';

/** Hooks */
import { useForm } from '../../hooks/useForm';

/** Functional Component */
export const TodoForm = ({ handleAddTodo }) => {

    const [ formValues, handleInputChange, reset ] = useForm({
        description: ''
    });

    // console.log( formValues );

    const handleSubmit = event => {
        event.preventDefault();

        if( formValues.description.trim().length <= 1 ) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            description: formValues.description,
            done: false
        }

        handleAddTodo( newTodo  );
        reset();

    }

    return (
        <form
            onSubmit={ handleSubmit }
        >

            <input 
                type="text"
                name="description"
                placeholder="DescripciÃ³n Tarea"
                autoComplete="off"
                className="form-control"
                value={ formValues.description }
                onChange={ handleInputChange }
            />
            <button 
                type="submit"
                className="btn btn-outline-primary btn-block mt-2 mb-2"
            >
                Agregar
            </button>

        </form>
    )
}
