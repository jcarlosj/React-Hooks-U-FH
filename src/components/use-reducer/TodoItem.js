import React from 'react';

export const TodoItem = ({ idx, todo, handleToggle, handleDelete }) => {
    return (
        <li 
            key={ todo.id }
            className="list-group-item d-flex justify-content-between align-items-center"
        >
            <p 
                className={ `task-description ${ todo.done && 'complete' }`}
                onClick={ () => handleToggle( todo.id ) }
            >{ idx + 1 }. { todo.description }</p>
            <button 
                className="btn btn-danger"
                onClick={ () => handleDelete( todo.id ) }
            >Borrar</button>
        </li>
    )
}
