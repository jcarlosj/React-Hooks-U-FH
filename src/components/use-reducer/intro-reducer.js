/*- 1 -- Estado inicial ----------------------------*/

/** Estado Inicial */
const initialState = [{
    id: 1,
    todo: 'Aprender Angular',
    done: false
}];

/** Define funcion pura como reducer (controlar acciones) */
const todoReducer = ( state = initialState, action ) => {

    if( action?.type === 'ADD' ){
        return [ ...state, action.payload ];
    }

    return state;
}

/** Llama el Reducer */
let todos = todoReducer();

/*- 2 - Insertar nuevo registro ----------------------------*/

/** Nuevo registro */
const newTodo = {
    id: 2,
    todo: 'Aprender Vue',
    done: false
};

/** Define la accion */
const addAction = {
    type: 'ADD',
    payload: newTodo
}

/** Llama el Reducer */
todos = todoReducer( todos, addAction );

/*- 1 - Nuestra registros ----------------------------*/
console.log( todos );