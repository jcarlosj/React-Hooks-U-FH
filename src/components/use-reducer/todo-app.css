p.task-description {
    cursor: pointer;
    margin: 0;
    padding: .375rem .75rem;
}

.complete {
    text-decoration: line-through;
} 