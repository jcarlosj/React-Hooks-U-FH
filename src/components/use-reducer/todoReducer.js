/** Define funcion pura como reducer (controlar acciones) */
export const todoReducer = ( state = [], action ) => {

    switch ( action.type ) {
        case 'ADD':
            return [ ...state, action.payload ];
        case 'DELETE':
            return state.filter( task => task.id !== action.payload );
        case 'TOGGLE': 
            return state.map( task =>       //  Return Implicito
                ( task.id === action.payload )
                    ?   { ...task, done: ! task.done } 
                    :   task
            );
        case 'TOGGLE-OLD':
            return state.map( task => {

                if( task.id === action.payload ) {
                    return {
                        ...task,
                        done: ! task.done
                    }
                } 

                return task;
            });

        default:
            return state;
    }

}