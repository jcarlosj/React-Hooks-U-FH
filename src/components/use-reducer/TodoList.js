import React from 'react';

/** Component */
import { TodoItem } from './TodoItem';

/** Functional Component */
export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul className="list-group list-group-flush">
            {    todos.map( ( todo, idx ) => (
                    <TodoItem 
                        key={ idx }
                        idx={ idx } 
                        todo={ todo }
                        handleToggle={ handleToggle }
                        handleDelete={ handleDelete }
                    />
                ))
            }
        </ul>
    )
}
