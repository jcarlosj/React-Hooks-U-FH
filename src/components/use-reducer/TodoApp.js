import React, { useEffect, useReducer } from 'react';

/** Hooks */
import { useForm } from '../../hooks/useForm';

/** Reducer */
import { todoReducer } from './todoReducer';

import './todo-app.css';

/** Establecera el initialState del Reducer */
const init = () => {
    return [{
        id: new Date().getTime(),
        description: 'Aprender Angular',
        done: false
    }];
}

/** Functional Component */
export const TodoApp = () => {

    const 
        [ todos, dispatch ] = useReducer( todoReducer, [], init ),      //  Establece State usando Inicialización diferida del Reducer. permite extraer la lógica para calcular el estado inicial fuera del reductor. También es útil para reiniciar el estado luego en respuesta a una acción. 
        [ formValues, handleInputChange, reset ] = useForm({
            description: ''
        });

    console.log( formValues );
    
    const handleSubmit = event => {
        event.preventDefault();

        if( formValues.description.trim().length <= 1 ) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            description: formValues.description,
            done: false
        }

        const action = {
            type: 'ADD',
            payload: newTodo
        }

        dispatch( action );
        reset();

    }

    return (
        <section className="container mt-5">

            <h1>To Do ( { todos.length } ) </h1>  
            <p>Usando el Hook useReducer</p>  
            <hr /> 

            <div className="row">
                <div className="col-5">

                    <form
                        onSubmit={ handleSubmit }
                    >

                        <input 
                            type="text"
                            name="description"
                            placeholder="Descripción Tarea"
                            autoComplete="off"
                            className="form-control"
                            value={ formValues.description }
                            onChange={ handleInputChange }
                        />
                        <button 
                            type="submit"
                            className="btn btn-outline-primary btn-block mt-2 mb-2"
                        >
                            Agregar
                        </button>

                    </form>

                </div>
                <div className="col-7">

                    { todos.length > 0 && (
                        <>
                            <h3>Lista de tareas</h3>
                            <hr />
                        </>
                    )}

                    <ul className="list-group list-group-flush">
                        {    todos.map( ( todo, idx ) => (
                                    <li 
                                        key={ todo.id }
                                        className="list-group-item d-inline-flex align-items-center"
                                    >
                                        <p className="task-description">{ idx + 1 }. { todo.description }</p>
                                        <button className="btn btn-danger">Borrar</button>
                                    </li>
                                ))
                        }
                    </ul>

                </div>
            </div>

        </section>
    )
}
