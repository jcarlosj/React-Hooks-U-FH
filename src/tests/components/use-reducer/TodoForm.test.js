import '@testing-library/jest-dom';
import { shallow } from 'enzyme';

import { TodoForm } from '../../../components/use-reducer/TodoForm';

describe( '<TodoForm />', () => {

    const
        handleAddTodo = jest.fn(), 
        wrapper = shallow(
            <TodoForm handleAddTodo={ handleAddTodo } />
        );

    test( 'debe desplegar correctamente', () => {

        // console.log( wrapper.html() );
        expect( wrapper ).toMatchSnapshot();

    } );

} );