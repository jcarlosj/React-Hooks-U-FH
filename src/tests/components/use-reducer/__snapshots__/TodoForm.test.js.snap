// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoForm /> debe desplegar correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    className="form-control"
    name="description"
    onChange={[Function]}
    placeholder="DescripciÃ³n Tarea"
    type="text"
    value=""
  />
  <button
    className="btn btn-outline-primary btn-block mt-2 mb-2"
    type="submit"
  >
    Agregar
  </button>
</form>
`;
